Update from Mongo 4.4 to Mongo 7 in CI

cognifloyd opened the pull request:
Switches from Mongo 4.4 to Mongo 7.0 using work by @guzzijones cherry-picked from #6236.

Also updates the redis container definition in the orquesta workflow to be consistent with ci workflow changes added in #6245.

nzlosh commented on st2common/st2common/models/db/__init__.py:
```diff
@@ -444,34 +453,37 @@ def _get_ssl_kwargs(
     }
+    # TODO: replace ssl_keyfile and ssl_certfile with tlsCertificateFile per pymongo:
+    #   > Instead of using ssl_certfile and ssl_keyfile to specify the certificate
+    #   > and private key files respectively, use tlsCertificateKeyFile to pass a
+    #   > single file containing both the client certificate and the private key.
+    # The tlsCertificateFile switch will be user-facing as files must be combined.
     if ssl_keyfile:
-        ssl_kwargs["ssl"] = True
+        ssl_kwargs["tls"] = True
         ssl_kwargs["ssl_keyfile"] = ssl_keyfile
     if ssl_certfile:
-        ssl_kwargs["ssl"] = True
+        ssl_kwargs["tls"] = True
         ssl_kwargs["ssl_certfile"] = ssl_certfile
```
Upgrade notes should explain the pymongo requirement and instruct the user to `cat client.key client.crt > client_combined.pem` and update `st2.conf` to remove the old key/cert `ssl_keyfile` and `ssl_certfile` entries and replace them with a single entry:

```tlscertificatekeyfile = client_combined.pem```

If the code allows it, we could check if `st2.conf` contains `ssl_keyfile` or `ssl_certfile` and raise an error to prevent StackStorm from starting, but the current behaviour is to simply ignore unknown key-pairs in the configuration file, so it may be non-trivial to implement.

amanda11 replied:
It may be easier as well for rolling upgrades to allow them to add in the new parameters and keep the old ones, that way the new ones will only be read by the instances that have upgraded and ignore the others.
So I think we just need to make sure it goes into the upgrade notes as a breaking change

cognifloyd replied:
Makes sense. Should we add all the other `tls*` variants to st2.conf as well?

Maybe there is a way to mark an option deprecated, which would result in a non-fatal warning...

cognifloyd replied:
OK. I renamed the ssl vars. Please take another look.

cognifloyd commented:
This has become so huge, I think it might need to be split into 3 PRs:
- migrate the SSL -> TLS database option names 
- update mongoengine and pymongo 
- move CI to mongo 7

If anyone wants to review this as is, go for it. I won't start splitting this up until Monday or Tuesday (Sept 23-24).

nzlosh commented on conf/st2.conf.sample:
```diff
@@ -155,10 +151,22 @@ ssl_certfile = None
+# Create the connection to mongodb using TLS.
+tls = False
+# Specifies whether MongoDB is allowed to pass an invalid certificate. This defaults to False to have security by default. Only temporarily set to True if you need to debug the connection.
+tlsAllowInvalidCertificates = None
```
I prefer remaining consistent with the rest of the `st2.conf` variable naming convention and use snake case as opposed to MongoDB's camel case.

nzlosh commented on st2common/st2common/config.py:
```diff
@@ -179,16 +179,24 @@ def register_opts(ignore_errors=False):
 
     do_register_opts(webui_opts, "webui", ignore_errors)
 
+    # NOTE: These options, where applicable, try to use the standard mongo name, even though
```
Are we saying the reason for camel case in `st2.conf` is so debug presents the same name as the mongo documentation?

cognifloyd replied:
Yes, though searching for `tlsFoo` vs `tls_foo` is probably not a big deal. So, I'm flexible here.

Pymongo changed their arg names to match a standard for mongo libraries across all languages:
- https://pymongo.readthedocs.io/en/stable/migrate-to-pymongo4.html#renamed-uri-options
- https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.md

I debated whether to follow the same standard or use snake_case. I decided to try following it and see what feedback TSC members give. Switching option names from camelCase to snake_case should be fairly simple since I kept snake_case within the code.

Does anyone else have a preference here?

guzzijones commented:
I would rather not split this out.  we can do the pymongo update in a separate pr in the future to avoid the debug log flood.  I am guessing we can filter those out with pytest anyway.

cognifloyd commented:
I moved this back to draft as I extracted the `[database].ssl` -> `[database].tls` option renaming in #6250. In that PR, I used `snake_case` for the options instead of the `camelCase` names in this PR. I also updated the `st2.conf.sample` generation to better report deprecations and use newer `oslo_config` features.

cognifloyd commented:
I just rebased this on master now that #6250 and #6252 have been merged. This PR is now focused on Makefile / CI updates to support MongoDB 7. So, this is ready for review.
